Regression(251783@main) Back navigation after browsing source tree on GitHub unexpectedly skips history items

cdumez opened the pull request:
#### 048c68aca363063546a4efb9e64f210ef706d8d5
<pre>
Regression(251783@main) Back navigation after browsing source tree on GitHub unexpectedly skips history items
<a href="https://bugs.webkit.org/show_bug.cgi?id=242947">https://bugs.webkit.org/show_bug.cgi?id=242947</a>
&lt;rdar://97031026&gt;

Reviewed by Geoffrey Garen.

Source tree navigations on github cause a history.pushState() to occur asynchronously,
causing us to lose track of the fact that the JS-triggered navigation was indeed the
result of user interaction. As a result, we would unexpectedly skip the created
HistoryItems on back navigation.

To address the issue, we relax our policy and mark HistoryItem as not having a user
gesture if there was no user gesture within the last second.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::hasRecentUserInteractionForNavigationFromJS const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::continueFragmentScrollAfterNavigationPolicy):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::FrameLoader::HistoryController::pushState):
* Tools/TestWebKitAPI/Tests/WebKit/WKBackForwardList.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/252649@main">https://commits.webkit.org/252649@main</a>
</pre>

geoffreygaren reviewed (approved):
r=me

geoffreygaren reviewed (commented):
Does the code prevent a website from adding multiple history entries from a single user gesture? I think there must be some code for that somewhere, but I can't see where.

cdumez committed:
Regression(251783@main) Back navigation after browsing source tree on GitHub unexpectedly skips history items
https://bugs.webkit.org/show_bug.cgi?id=242947
<rdar://97031026>

Reviewed by Geoffrey Garen.

Source tree navigations on github cause a history.pushState() to occur asynchronously,
causing us to lose track of the fact that the JS-triggered navigation was indeed the
result of user interaction. As a result, we would unexpectedly skip the created
HistoryItems on back navigation.

To address the issue, we relax our policy and mark HistoryItem as not having a user
gesture if there was no user gesture within the last second.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::hasRecentUserInteractionForNavigationFromJS const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::continueFragmentScrollAfterNavigationPolicy):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::FrameLoader::HistoryController::pushState):
* Tools/TestWebKitAPI/Tests/WebKit/WKBackForwardList.mm:
(TEST):

Canonical link: https://commits.webkit.org/252649@main
